### PR TITLE
refactor: add token standard folders & cpy erc20 permit2

### PR DESCRIPTION
**Pull request by snreynolds**
file structure is essentially
src/
-- ERC20/
---- libraries/
---- interfaces/
---- Permit2.sol etc...
-- ERC721/
---- libraries/
---- interfaces/
---- Permit2.sol etc...

There _might_ be some shared libraries but I can pull those out at the end into a shared folder although I don't think a lot will be shared since the sigs will now sign tokenId and all the transfers will have tokenId. Thoughts?

### CODE REVIEW

File: src/ERC721/Permit2.sol [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
```
Comment (marktoda) [inline src/ERC721/Permit2.sol:2]: Consider naming these concrete contracts `ERC20Permit2` and `ERC721Permit2` for clarity

**Review by marktoda — COMMENTED** [review]
What is your plan for the test folder? I would consider trying to work out a framework to have a (at least partially) shared test suite, to ensure even coverage. I.e. shared test suite for nonce / witness management  would be great since that logic should be exactly same between the two. Maybe base test class with type-specific overrides for `permit(nonce)` or something

File: src/ERC721/Permit2.sol (hunk 1 above)
Comment (snreynolds) [reply src/ERC721/Permit2.sol:2]: Yeah, agree - will make that change. I wont change the ERC20 permit2 naming but for future token standards I think I'll rename to Permit2_ERC721 and Permit2_ERC1155 etc..

**Comment by snreynolds** [comment]
> What is your plan for the test folder? I would consider trying to work out a framework to have a (at least partially) shared test suite, to ensure even coverage. I.e. shared test suite for nonce / witness management would be great since that logic should be exactly same between the two. Maybe base test class with type-specific overrides for `permit(nonce)` or something

Yeah good point, I think the unit tests can definitely be shared (nonce bitmap tests and compact sig tests but probably not the packed allowance unit tests since the naming will be different, amount -> tokenId and there will be new behavior there). Will be tougher to share the general allowance and signature transfer tests bc the test tokens are different but maybe there is some abstraction we can do? lmk if you have ideas but I'll probably just have to play around when i get to testing

**Pull request merged**